pkg/cmd/pulumi/errors.go

```diff
@@ -11,28 +11,51 @@ import (
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/result"
+	"github.com/spf13/cobra"
 )
 
-// PrintEngineResult optionally provides a place for the CLI to provide human-friendly error
-// messages for messages that can happen during normal engine operation.
-func PrintEngineResult(err error) error {
-	// If we had no actual result, or the result was a request to 'Bail', then we have nothing to
-	// actually print to the user.
-	if err == nil || result.IsBail(err) {
-		return err
+// runCmdFunc wraps cmdutil.RunFunc. While cmdutil.RunFunc provides a standard
+// wrapper for dealing with and logging errors before exiting with an
+// appropriate error code, runCmdFunc extends this with additional error
+// handling specific to the Pulumi CLI. This includes e.g. specific and more
+// helpful messages in the case of decryption or snapshot integrity errors.
+func runCmdFunc(
+	run func(cmd *cobra.Command, args []string) error,
+) func(cmd *cobra.Command, args []string) {
+	return cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+		err := run(cmd, args)
+		return processCmdErrors(err)
+	})
+}
+
+// Processes errors that may be returned from commands, providing a central
+// location to insert more human-friendly messages when certain errors occur, or
+// to perform other type-specific handling.
+func processCmdErrors(err error) error {
+	// If no error occurred, we have nothing to do.
+	if err == nil {
+		return nil
 	}
 
-	switch e := err.(type) {
-	case engine.DecryptError:
-		printDecryptError(e)
-		// We have printed the error already.  Should just bail at this point.
+	// If the error is a "bail" (that is, some expected error flow), then a
+	// diagnostic or message will already have been reported and we can swallow
+	// it.
+	if result.IsBail(err) {
 		return nil
```

Comment:
Hrm if we swallow it here, the error will not be bubbled up to `main.go`, and we'll exit with exit code 0, where as currently we would `exit 1`.  Same below for the decrypt error.  I think we can set `cmd.SilenceErrors` in `runCmdFunc` instead to make cobra not print the error (which I presume is the reason for swallowing it), but still return the error to get the right exit codes.

Reply:
Absolute hawk-eye review! Great catch. I've amended this so that:

* We match existing behaviour for `Bail`s right now, which is return `err` and not `nil` (that was a bug on my part; fixed)
* For snapshot integrity errors, we wrap the error in a `Bail` so that `RunFunc` will still exit with a non-zero exit code, but it won't reprint the (non-panic-bannered) error.

I think this preserves existing behaviour correctly and adds in the new behaviour in a satisfactory manner. There is an open question of whether the existing behaviour for `DecryptError` (which will result in zero exit codes) is correct -- I think not, but will open a separate PR for that and discuss with the wider team.